### PR TITLE
DMP-3989 - Introduce and use manual deletion feature flag

### PR DESCRIPTION
**Pull request by Ben-Edwards-cgi**
### Change description ###
*Add feature flag*

If this flag has not already been created, add the new config to DARTS API application.yaml, to act as a feature flag for manual deletion
{code:java}
darts:
  manual-deletion:
    enabled: ${MANUAL_DELETION_ENABLED:false} {code}
Introduce the env var "MANUAL_DELETION_ENABLED" in flux for the darts-api and darts-portal services staging environment, set to true. Ensure it is not set or set to false for all other environments.

*Amend hidden reasons endpoint*

Amend the /admin/hidden-reasons endpoint to only return the reasons with "marked_for_deletion" = true if manual deletion is enabled via the feature flag.

Ignore the "display_state" value, as the feature flag is being used as an override for this in order to control manual deletion from one place.

*Amend media/transcription document hide endpoints*

Amend the POST /admin/medias/\{media_id}/hide and POST /admin/transcription-documents/\{transcription_document_id}/hide endpoints to only allow deletion if manual deletion is enabled via the feature flag.

DMP-2891 managed the transcription document hide endpoint via a validate method, this check will need to be conditioned using the feature flag. Ensure the media hide endpoint behave in a similar way.

*Amend media/transcription document marked-for-deletion endpoints*

Only allow these endpoints to be used if manual deletion is enabled via the feature flag, otherwise return 501 not implemented.

GET /admin/medias/marked-for-deletion
GET /admin/transcription-documents/marked-for-deletion

*Amend the media/transcription document approve-deletion endpoints*

Only allow these endpoints to be used if manual deletion is enabled via the feature flag, otherwise return 501 not implemented.

POST /admin/medias/\{media_id}/approve-deletion
POST /admin/transcription-documents/\{transcription_document_id}/approve-deletion
h3.  
h3.  
h3. *Acceptance Criteria*

*AC1 - feature enabled - hidden reasons*

GIVEN manual deletion is enabled by the feature flag
WHEN the GET /admin/hidden-reasons endpoint is requested
THEN the response includes all reasons, including those which mark as deleted

*AC2 - feature not enabled - hidden reasons*

GIVEN manual deletion is not enabled by the feature flag
WHEN the GET /admin/hidden-reasons endpoint is requested
THEN the response does not including reasons which mark as deleted

*AC3 - feature enabled - media/transcription document hide endpoint*

GIVEN manual deletion is enabled by the feature flag
WHEN the POST /admin/medias/\{media_id}/hide OR POST /admin/transcription-documents/\{transcription_document_id}/hide endpoint is requested
THEN the response is 200 and an entry is made in the object_admin_action table

*AC4 - feature not enabled - media/transcription document hide endpoint*

GIVEN manual deletion is enabled by the feature flag
WHEN the POST /admin/medias/\{media_id}/hide OR POST /admin/transcription-documents/\{transcription_document_id}/hide endpoint is requested
THEN the response is 501

*AC5 - feature enabled - media/transcription document marked-for-deletion endpoint*

GIVEN manual deletion is enabled by the feature flag
WHEN the GET /admin/medias/marked-for-deletion OR GET /admin/transcription-documents/marked-for-deletion endpoint is requested
THEN the response is 200 with a list of items that are marked for deletion

*AC6 - feature not enabled - media/transcription document marked-for-deletion endpoint*

GIVEN manual deletion is enabled by the feature flag
WHEN the GET /admin/medias/marked-for-deletion OR GET /admin/transcription-documents/marked-for-deletion endpoint is requested
THEN the response is 501

*AC7 - feature enabled - media/transcription document approve-deletion endpoint*

GIVEN manual deletion is enabled by the feature flag
WHEN the POST /admin/media/approve-deletion OR POST /admin/transcription-documents/approve-deletion endpoint is requested
THEN the response is 200 and the item is marked for deletion

*AC8 - feature not enabled - media/transcription document approve-deletion endpoint*

GIVEN manual deletion is enabled by the feature flag
WHEN the POST /admin/media/approve-deletion OR POST /admin/transcription-documents/approve-deletion endpoint is requested
THEN the response is 501

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/darts/common/exception/DartsApiException.java [hunk 1]
```diff
@@ -59,4 +62,25 @@ public DartsApiException(DartsApiError error, String detail, Throwable throwable
         this.detail = detail;
     }
 
+    @Getter
+    @RequiredArgsConstructor
+    public enum DartsApiErrorCommon implements DartsApiError {
+        FEATURE_FLAG_NOT_ENABLED(
+            "FEATURE_FLAG_NOT_ENABLED",
+            HttpStatus.NOT_IMPLEMENTED,
+            AddAudioTitleErrors.USER_CANT_APPROVE_THEIR_OWN_DELETION.getValue()
```
Comment (davet1985) [inline src/main/java/uk/gov/hmcts/darts/common/exception/DartsApiException.java:71]: This doesn't look like the right thing to use here

File: src/main/java/uk/gov/hmcts/darts/transcriptions/service/impl/AdminTranscriptionServiceImpl.java [hunk 2]
```diff
@@ -161,6 +168,9 @@ public GetTranscriptionDocumentByIdResponse getTranscriptionDocumentById(Integer
     @Transactional
     public TranscriptionDocumentHideResponse hideOrShowTranscriptionDocumentById(Integer transcriptionDocumentId,
                                                                                  TranscriptionDocumentHideRequest transcriptionDocumentHideRequest) {
+        if (!this.isManualDeletionEnabled()) {
+            throw new DartsApiException(DartsApiException.DartsApiErrorCommon.FEATURE_FLAG_NOT_ENABLED);
+        }
```
Comment (davet1985) [inline src/main/java/uk/gov/hmcts/darts/transcriptions/service/impl/AdminTranscriptionServiceImpl.java:173]: The check here needs to be more subtle. The `reason_id` in the payload needs to be checked against the `marked_for_deletion` flag on the `object_hidden_reason` table and if it's true then the flag must be enabled to proceed.

This is currently handled in the `TranscriptionDocumentHideOrShowValidator.validate` method, please look there and adjust appropriately.

File: src/main/java/uk/gov/hmcts/darts/audio/service/impl/MediaRequestServiceImpl.java [hunk 3]
```diff
@@ -541,6 +548,9 @@ private void auditOwnerChange(MediaPatchRequest request, MediaRequestEntity medi
     @Override
     @Transactional
     public MediaHideResponse adminHideOrShowMediaById(Integer mediaId, MediaHideRequest mediaHideRequest) {
+        if (!this.isManualDeletionEnabled()) {
+            throw new DartsApiException(DartsApiException.DartsApiErrorCommon.FEATURE_FLAG_NOT_ENABLED);
+        }
```
Comment (davet1985) [inline src/main/java/uk/gov/hmcts/darts/audio/service/impl/MediaRequestServiceImpl.java:553]: This need to behave the same as the transcription document hide endpoint, checking the `marked_for_deletion` flag for the associated `reason_id`.

**Review by davet1985 — CHANGES_REQUESTED** [review]
Need to amend the logic on the `/hide` endpoint please.

File: src/main/java/uk/gov/hmcts/darts/common/exception/DartsApiException.java (hunk 1 above)
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/common/exception/DartsApiException.java:71]: Good spot updated

File: src/main/java/uk/gov/hmcts/darts/transcriptions/service/impl/AdminTranscriptionServiceImpl.java (hunk 2 above)
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/transcriptions/service/impl/AdminTranscriptionServiceImpl.java]: Updated

File: src/main/java/uk/gov/hmcts/darts/audio/service/impl/MediaRequestServiceImpl.java (hunk 3 above)
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/audio/service/impl/MediaRequestServiceImpl.java]: Updated

File: src/test/java/uk/gov/hmcts/darts/audio/validation/MediaHideOrShowValidatorTest.java [hunk 4]
```diff
@@ -155,4 +162,60 @@ void failWhenHideWithActionRequestAndWithoutCorrectReason() {
...
+        IdRequest<MediaHideRequest> mediaHideRequestIdRequest = new
+            IdRequest<>(mediaHideRequest, mediaId);
+
+        mediaHideOrShowValidator.validate(mediaHideRequestIdRequest);
+
+        Mockito.verify(mediaIdValidator, times(1)).validate(mediaId);
+    }
+
+    @Test
+    void successfullyHideWithMarkedForDeletionActionRequestAndFlagDisabled() {
```
[the remark this answers is not in the document]
Comment (Ben-Edwards-cgi) [reply src/test/java/uk/gov/hmcts/darts/audio/validation/MediaHideOrShowValidatorTest.java:194]: Updated

File: src/test/java/uk/gov/hmcts/darts/transcriptions/validator/TranscriptionDocumentHideOrShowValidatorTest.java [hunk 5]
```diff
@@ -186,4 +194,33 @@ void failWhenHideWithActionRequestAndWithReasonMarkedForDeletion() {
 
         Mockito.verify(transcriptionDocumentIdValidator, times(1)).validate(documentId);
     }
+
+    @Test
+    void failWhenHideWithActionRequestAndWithReasonMarkedForDeletionButManualDeletionIsDisabled() {
```
[the remark this answers is not in the document]
Comment (Ben-Edwards-cgi) [reply src/test/java/uk/gov/hmcts/darts/transcriptions/validator/TranscriptionDocumentHideOrShowValidatorTest.java:199]: Added

File: src/main/java/uk/gov/hmcts/darts/transcriptions/validator/TranscriptionDocumentHideOrShowValidator.java [hunk 6]
```diff
@@ -50,8 +57,12 @@ public void validate(IdRequest<TranscriptionDocumentHideRequest> request) {
             } else {
                 ObjectHiddenReasonEntity objectHiddenReasonEntity = optionalObjectHiddenReasonEntity.get();
                 if (objectHiddenReasonEntity.isMarkedForDeletion()) {
-                    throw new DartsApiException(TranscriptionApiError
-                                                    .TRANSCRIPTION_DOCUMENT_REASON_IS_MARKED_FOR_DELETION);
+                    if (isManualDeletionEnabled()) {
+                        throw new DartsApiException(TranscriptionApiError
```
[the remark this answers is not in the document]
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/transcriptions/validator/TranscriptionDocumentHideOrShowValidator.java:61]: There is existing logic in the application here where if the objectHiddenReasonEntity was marked for deletion the below error would be thrown: TranscriptionApiError.TRANSCRIPTION_DOCUMENT_REASON_IS_MARKED_FOR_DELETION this is to accommodate the new and old exceptions. 

Although arguably we do not need the new exception if it is already rejected by the old

**Review by davet1985 — APPROVED** [review]
LGTM

**Pull request merged**